### PR TITLE
#103: CDC should purge Eventuate Tram message and received_messages t…

### PR DESCRIPTION
**Pull request by dartartem**
…ables. Created message cleaner. Partially covered by tests.

### CODE REVIEW

File: eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/MessageCleaner.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+import javax.sql.DataSource;
+import java.util.Timer;
+import java.util.TimerTask;
+
+public class MessageCleaner {
+
+  private EventuateSqlDialect eventuateSqlDialect;
+  private DataSource dataSource;
+  private EventuateSchema eventuateSchema;
+  private Boolean purgeMessagesEnabled;
```
Comment (cer) [inline eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/MessageCleaner.java:16]: Why is this Boolean? Can it be nullable?

File: eventuate-local-java-cdc-connector-unified/src/main/java/io/eventuate/local/unified/cdc/pipeline/common/configuration/CdcDefaultPipelinePropertiesConfiguration.java [hunk 2]
```diff
@@ -29,6 +29,11 @@ public CdcPipelineProperties cdcPipelineProperties(DefaultSourceTableNameResolve
             .ofNullable(eventuateConfigurationProperties.getSourceTableName())
             .orElse(sourceTableNameResolver.resolve("default")));
 
+    cdcPipelineProperties.setPurgeMessagesEnabled(eventuateConfigurationProperties.getPurgeMessagesEnabled());
```
Comment (cer) [inline eventuate-local-java-cdc-connector-unified/src/main/java/io/eventuate/local/unified/cdc/pipeline/common/configuration/CdcDefaultPipelinePropertiesConfiguration.java:32]: Why?
A cleaner should know about a pipeline. But a pipeline should not know about a cleaner.

File: eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/MessageCleaner.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  private int purgeIntervalInSeconds;
+
+  private Timer timer;
+  private JdbcTemplate jdbcTemplate;
+
+
+  public MessageCleaner(EventuateSqlDialect eventuateSqlDialect,
+                        DataSource dataSource,
+                        EventuateSchema eventuateSchema,
+                        Boolean purgeMessagesEnabled,
```
Comment (cer) [inline eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/MessageCleaner.java:30]: This pattern of fields is repeated:

```
Boolean purgeMessagesEnabled,
                        int purgeMessagesMaxAgeInSeconds,
                        Boolean purgeReceivedMessagesEnabled,
                        int purgeReceivedMessagesMaxAgeInSeconds,
                        int purgeIntervalInSeconds
```

There should be a class CleanerConfiguration? that is passed around.

File: eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/MessageCleaner.java (hunk 1 above)
Comment (dartartem) [reply eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/MessageCleaner.java:16]: Somehow I missed it, it should be boolean. I will review everywhere.

File: eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/EventuateConfigurationProperties.java [hunk 4]
```diff
@@ -79,6 +81,18 @@
   @Value("${eventuate.cdc.kafka.batch.processing.max.batch.size:#{1000000}}")
   private int maxBatchSize;
 
+  @Value("${purge.message.enabled:#{false}}")
+  private Boolean purgeMessagesEnabled;
```
Comment (cer) [inline eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/EventuateConfigurationProperties.java:85]: Is this nullable? Why `Boolean`?

**Review by cer — CHANGES_REQUESTED** [review]

File: eventuate-local-java-cdc-connector-unified/src/main/java/io/eventuate/local/unified/cdc/pipeline/common/configuration/CdcDefaultPipelinePropertiesConfiguration.java (hunk 2 above)
Comment (dartartem) [reply eventuate-local-java-cdc-connector-unified/src/main/java/io/eventuate/local/unified/cdc/pipeline/common/configuration/CdcDefaultPipelinePropertiesConfiguration.java:32]: ok, I will fix

File: eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/MessageCleaner.java (hunk 3 above)
Comment (dartartem) [reply eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/MessageCleaner.java:30]: ok, I will add

File: eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/EventuateConfigurationProperties.java (hunk 4 above)
Comment (dartartem) [reply eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/EventuateConfigurationProperties.java:85]: no, I will change to boolean

File: eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/MessageCleaner.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+    if (purgeReceivedMessagesEnabled) {
+      cleanReceivedMessages();
+    }
+  }
+
+  private void cleanMessages() {
+    String table = eventuateSchema.qualifyTable("message");
+
+    String sql = "delete from " + table +
```
Comment (cer) [inline eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/MessageCleaner.java:79]: Please use String.format() + JDBC/SQL placeholders rather than String concatentation.

**Review by cer — CHANGES_REQUESTED** [review]

File: eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/MessageCleaner.java (hunk 5 above)
Comment (dartartem) [reply eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/MessageCleaner.java:79]: ok

File: eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/MessageCleanerPurgeProperties.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import org.apache.commons.lang.builder.ReflectionToStringBuilder;
+import org.springframework.util.Assert;
+
+import java.util.concurrent.TimeUnit;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class MessageCleanerPurgeProperties{
+  private boolean purgeMessagesEnabled = false;
```
Comment (cer) [inline eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/MessageCleanerPurgeProperties.java:11]: This could be `messagesEnabled` ?

File: eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/MessageCleanerPurgeProperties.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
+package io.eventuate.local.common;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import org.apache.commons.lang.builder.ReflectionToStringBuilder;
+import org.springframework.util.Assert;
+
+import java.util.concurrent.TimeUnit;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
```
Comment (cer) [inline eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/MessageCleanerPurgeProperties.java:9]: Why is this annotation needed?

File: eventuate-local-java-cdc-connector-mysql-binlog/src/test/java/io/eventuate/local/mysql/binlog/MysqlMessageCleanerTest.java [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+import io.eventuate.local.test.util.AbstractMessageCleanerTest;
+import org.junit.runner.RunWith;
+import org.springframework.boot.test.context.SpringBootTest;
+
+import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
+
+@RunWith(SpringJUnit4ClassRunner.class)
+@SpringBootTest(classes = AbstractMessageCleanerTest.Config.class)
+public class MysqlMessageCleanerTest extends AbstractMessageCleanerTest {
```
Comment (cer) [inline eventuate-local-java-cdc-connector-mysql-binlog/src/test/java/io/eventuate/local/mysql/binlog/MysqlMessageCleanerTest.java:11]: `Mysql` --> `MySql`

File: eventuate-local-java-cdc-connector-unified/src/main/java/io/eventuate/local/unified/cdc/pipeline/common/properties/CdcPipelineProperties.java [hunk 9]
```diff
@@ -53,4 +62,44 @@ public String getSourceTableName() {
   public void setSourceTableName(String sourceTableName) {
     this.sourceTableName = sourceTableName;
   }
+
+  public Boolean getPurgeMessagesEnabled() {
```
Comment (cer) [inline eventuate-local-java-cdc-connector-unified/src/main/java/io/eventuate/local/unified/cdc/pipeline/common/properties/CdcPipelineProperties.java:66]: Why are these here?
Cleaner is nothing to do with pipeline

**Review by cer — CHANGES_REQUESTED** [review]

File: eventuate-local-java-cdc-connector-unified/src/main/java/io/eventuate/local/unified/cdc/pipeline/common/properties/CdcPipelineProperties.java (hunk 9 above)
Comment (dartartem) [reply eventuate-local-java-cdc-connector-unified/src/main/java/io/eventuate/local/unified/cdc/pipeline/common/properties/CdcPipelineProperties.java:66]: yes, missed to remove here

File: eventuate-local-java-cdc-connector-mysql-binlog/src/test/java/io/eventuate/local/mysql/binlog/MysqlMessageCleanerTest.java (hunk 8 above)
Comment (dartartem) [reply eventuate-local-java-cdc-connector-mysql-binlog/src/test/java/io/eventuate/local/mysql/binlog/MysqlMessageCleanerTest.java:11]: ok

File: eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/MessageCleanerPurgeProperties.java (hunk 7 above)
Comment (dartartem) [reply eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/MessageCleanerPurgeProperties.java:9]: it should not be needed, I will remove

File: eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/MessageCleanerPurgeProperties.java (hunk 6 above)
Comment (dartartem) [reply eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/MessageCleanerPurgeProperties.java:11]: yes, but it makes sense only in context. I wanted to invent something that does not repeat 'purge' but also makes sense without outer object. But if you think that is fine, I will just remove 'purge'.

File: eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/MessageCleaner.java [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+  private EventuateSchema eventuateSchema;
+  private MessageCleanerPurgeProperties messageCleanerPurgeProperties;
+
+  private Timer timer;
+  private JdbcTemplate jdbcTemplate;
+
+  public MessageCleaner(EventuateSqlDialect eventuateSqlDialect,
+                        DataSource dataSource,
+                        EventuateSchema eventuateSchema,
+                        MessageCleanerPurgeProperties messageCleanerPurgeProperties) {
```
[the remark this answers is not in the document]
Comment (dartartem) [reply eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/MessageCleaner.java:23]: ok
